source/elements/oneMKL/source/domains/spblas/operations/spmm.rst

```diff
@@ -0,0 +1,393 @@
+       Produces deterministic results.
+     - | MKL: N/A
+       | cuSPARSE: ``CUSPARSE_SPMM_COO_ALG2``
+       | rocSPARSE: ``rocsparse_spmm_alg_coo_atomic``
+   * - ``coo_alg3``
+     - Should provide best performance for COO format and large ``nnz``.
+     - | MKL: N/A
+       | cuSPARSE: ``CUSPARSE_SPMM_COO_ALG3``
+       | rocSPARSE: ``rocsparse_spmm_alg_coo_segmented_atomic``
+   * - ``coo_alg4``
+     - Should provide best performance for COO format and row-major layout.
+     - | MKL: N/A
+       | cuSPARSE: ``CUSPARSE_SPMM_COO_ALG4``
+       | rocSPARSE: N/A
+   * - ``csr_alg1``
+     - Should provide best performance for CSR format and column-major layout.
+     - | MKL: N/A
+       | cuSPARSE: ``CUSPARSE_SPMM_CSR_ALG1``
+       | rocSPARSE: ``rocsparse_spmm_alg_csr``
+   * - ``csr_alg2``
+     - Should provide best performance for CSR format and row-major layout.
+     - | MKL: N/A
+       | cuSPARSE: ``CUSPARSE_SPMM_CSR_ALG2``
+       | rocSPARSE: ``rocsparse_spmm_alg_csr_row_split``
+   * - ``csr_alg3``
+     - Deterministic algorithm for CSR format.
+     - | MKL: N/A
+       | cuSPARSE: ``CUSPARSE_SPMM_CSR_ALG3``
+       | rocSPARSE: ``rocsparse_spmm_alg_csr_merge``
+
+.. _onemkl_sparse_spmm:
+
+spmm
+----
+
+.. rubric:: Syntax
+
+.. code-block:: cpp
+
+   namespace oneapi::mkl::sparse {
+
+       void spmm_buffer_size(
+           sycl::queue                                &queue,
+           oneapi::mkl::transpose                     opA,
+           oneapi::mkl::transpose                     opB,
+           void*                                      alpha,
+           oneapi::mkl::sparse::matrix_view           A_view,
+           oneapi::mkl::sparse::matrix_handle_t       A_handle,
+           oneapi::mkl::sparse::dense_matrix_handle_t B_handle,
+           void*                                      beta,
+           oneapi::mkl::sparse::dense_matrix_handle_t C_handle,
+           oneapi::mkl::sparse::spmm_alg              alg,
+           oneapi::mkl::sparse::spmm_descr_t          spmm_descr,
+           std::size_t                                &temp_buffer_size,
+       );
+
+       void spmm_optimize(
+           sycl::queue                                &queue,
+           oneapi::mkl::transpose                     opA,
+           oneapi::mkl::transpose                     opB,
+           void*                                      alpha,
+           oneapi::mkl::sparse::matrix_view           A_view,
+           oneapi::mkl::sparse::matrix_handle_t       A_handle,
+           oneapi::mkl::sparse::dense_matrix_handle_t B_handle,
+           void*                                      beta,
+           oneapi::mkl::sparse::dense_matrix_handle_t C_handle,
+           oneapi::mkl::sparse::spmm_alg              alg,
+           oneapi::mkl::sparse::spmm_descr_t          spmm_descr,
+           sycl::buffer<std::uint8_t, 1>              workspace,
```

Comment:
That's a very good point, I think the issue is the same with a `void*` USM pointer. I think it would be enough to document that the address must be aligned on the data type used. I'm not a fan of generic types but it makes things a lot easier with the design of oneMKL to be honest.

Reply:
Added the requirement for workspaces to be aligned in https://github.com/oneapi-src/oneAPI-spec/pull/522/commits/2545e7b8d69cc51f3d347a3070b4f2dbddbc97fa